### ISSUE TITLE
CVE-2019-10010: XSS Vulnerability

### ISSUE THREAD

**Issue by colinodell** [thephpleague/commonmark#353]
An XSS vulnerability (CVE-2019-10010) has been identified in all previous versions of this library (0.18.2 and below).

The issue has been fixed in [version 0.18.3](https://github.com/thephpleague/commonmark/releases/tag/0.18.3).  **All users should upgrade to [version 0.18.3](https://github.com/thephpleague/commonmark/releases/tag/0.18.3) immediately.** Additionally, if your application caches the resulting HTML, please purge and/or regenerate those caches.

## Description

In `league/commonmark` 0.18.2 and below, malicious users can insert double-encoded HTML entities into their Markdown like this:

```md
[XSS](javascript&amp;colon;alert%28&#039;XSS&#039;%29)
```

This library would (correctly) unescape the `&amp;` entity to `&` during the parsing step.  However, **the renderer step would fail to properly re-escape the resulting `&colon;` string**, thus producing the following malicious HTML output:

```html
<p><a href="javascript&colon;alert('XSS')">XSS</a></p>
```

Browsers would interpret `&colon;` as a `:` character and allow the JS to be executed when the link is clicked.

This vulnerability was present in the upstream library this project was forked from and therefore exists in all prior versions of `league/commonmark`.

## Solution

The new [0.18.3](https://github.com/thephpleague/commonmark/releases/tag/0.18.3) release mirrors [the fix made upstream](https://github.com/commonmark/commonmark.js/commit/c89b35c5fc99bdf1d2181f7f0c9fcb8a1abc27c8) - we no longer attempt to preserve entities when rendering HTML attributes like `href`, `src`, `title`, etc.

The `$preserveEntities` parameter of `Xml::escape()` is therefore no longer used internally, so it has been deprecated and marked for removal in the next major release (0.19.0).

## Credits

 - Sebastiaan Knijnenburg for identifying the issue
 - Ross Tuck for responsibly disclosing the issue
 - John MacFarlane for investigating it and implementing the upstream fix we mirrored here

**Comment by colinodell** [comment]
The issue has been fixed in [version 0.18.3](https://github.com/thephpleague/commonmark/releases/tag/0.18.3).  **All users should upgrade to [version 0.18.3](https://github.com/thephpleague/commonmark/releases/tag/0.18.3) immediately.** 

The original message above has also been updated with additional details.

**Comment by colinodell** [comment]
This vulnerability has been assigned CVE ID CVE-2019-10010
